File: ReverseRecursive.java
// Time Complexity : O(n)
// Space Complexity : O(n) //recurisve stack
// Did this code successfully run on Leetcode : yes
class Solution {
    public ListNode reverseList(ListNode head) {
        
        if(head==null || head.next==null) return head;
        ListNode reversed = reverseList(head.next);
        head.next.next= head;
        head.next = null;
        
        return reversed;
        
    }
 
}

File: reverse-recursive.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* prev;
public:
    ListNode* reverseList(ListNode* head) {
     
       
        return reverse(head);
        
    }
    
    ListNode* reverse(ListNode* head) {
        
        // base
      
           if(head == NULL || head->next==NULL) return head;
        
        ListNode* temp = reverse(head->next);
        head->next->next = head;
        cout<<"setting null: "<<head->val<<endl;
        head->next = NULL;
    
        return temp;
        
    }
};
